src/hotspot/share/opto/superword.cpp

```diff
@@ -1214,110 +1186,136 @@ bool SuperWord::follow_def_uses(Node_List* p) {
+void SuperWord::order_inputs_of_all_use_pairs_to_match_def_pair(Node* def1, Node* def2) {
+  assert(_pairset.has_pair(def1, def2), "(def1, def2) must be a pair");
 
-  if (s1->is_Store()) return;
+  if (def1->is_Store()) return;
 
   // reductions are always managed beforehand
-  if (is_marked_reduction(s1)) return;
+  if (is_marked_reduction(def1)) return;
 
-  for (DUIterator_Fast imax, i = s1->fast_outs(imax); i < imax; i++) {
-    Node* t1 = s1->fast_out(i);
+  for (DUIterator_Fast imax, i = def1->fast_outs(imax); i < imax; i++) {
+    Node* use1 = def1->fast_out(i);
 
     // Only allow operand swap on commuting operations
-    if (!t1->is_Add() && !t1->is_Mul() && !VectorNode::is_muladds2i(t1)) {
+    if (!use1->is_Add() && !use1->is_Mul() && !VectorNode::is_muladds2i(use1)) {
       break;
     }
 
-    // Now find t1's packset
-    Node_List* p2 = nullptr;
-    for (int j = 0; j < _packset.length(); j++) {
-      p2 = _packset.at(j);
-      Node* first = p2->at(0);
-      if (t1 == first) {
-        break;
-      }
-      p2 = nullptr;
-    }
-    // Arrange all sub components by the major component
-    if (p2 != nullptr) {
-      for (uint j = 1; j < p->size(); j++) {
-        Node* d1 = p->at(j);
-        Node* u1 = p2->at(j);
-        opnd_positions_match(s1, t1, d1, u1);
-      }
-    }
+    // Find pair (use1, use2)
+    Node* use2 = _pairset.get_right_or_null_for(use1);
+    if (use2 == nullptr) { break; }
+
+    order_inputs_of_uses_to_match_def_pair(def1, def2, use1, use2);
   }
 }
 
-//---------------------------opnd_positions_match-------------------------
-// Is the use of d1 in u1 at the same operand position as d2 in u2?
-bool SuperWord::opnd_positions_match(Node* d1, Node* u1, Node* d2, Node* u2) {
-  // check reductions to see if they are marshalled to represent the reduction
-  // operator in a specified opnd
-  if (is_marked_reduction(u1) && is_marked_reduction(u2)) {
-    // ensure reductions have phis and reduction definitions feeding the 1st operand
-    Node* first = u1->in(2);
+// For a def-pair (def1. def2), and their use-nodes (use1, use2):
+// ensure that the input order of (use1, use2) matches the order of (def1, def2).
+//
+// We have different cases:
+//
+// 1. Reduction (use1, use2): must always reduce left-to-right. Make sure that we have pattern:
+//
+//    phi/reduction x1  phi/reduction x2                    phi/reduction x1
+//                | |               | |    and hopefully:               | |
+//                use1              use2                                use1 x2
+//                                                                         | |
+//                                                                         use2
+//
+// 2: Inputs of (use1, use2) already match (def1, def2), i.e. for all input indices i:
+//
+//    use1->in(i) == def1 || use2->in(def2)   ->    use1->in(i) == def1 && use2->in(def2)
+//
+// 3: Add/Mul (use1, use2): we can try to swap edges:
```

Comment:
Is it not required for other nodes?

Reply:
We can only do it for associative nodes like `Mul / Add`, and of course all the nodes that are subclasses of those (e.g. Max, Min, Or, etc). I can improve the comment, since that may not be immediately clear to the reader.

Reply:
Thanks for the explanation. Yes, maybe you can clarify the comment :-)